dropshot/src/extractor/mod.rs

```diff
@@ -23,6 +26,70 @@ pub use path::Path;
 
 mod query;
 pub use query::Query;
+pub use query::QueryError;
 
 mod raw_request;
 pub use raw_request::RawRequest;
+
+use crate::error::HttpError;
+pub use crate::http_util::PathError;
+
+/// Errors returned by extractors.
+///
+/// Because extractors can be combined (a tuple of types which are extractors is
+/// itself an extractor), all extractors must return the same error type. Thus,
+/// this is an enum of all the possible errors that can be returned by any
+/// extractor.
```

Comment:
I'm not sure about this approach here.  These are just the extractors that Dropshot provides.  But consumers can define their own, too.

Also, it seems like a lot of the actual error cases probably overlap (e.g., you can have "invalid input" in a typed body, path param, query param, etc. and it's not clear a consumer should have to handle these all separately).

nit: it seems like this belongs in dropshot/src/extractor/common.rs (or a new file) -- mod.rs is just a thing that includes other files and "common" is where we put the common stuff here.

Reply:
> I'm not sure about this approach here. These are just the extractors that Dropshot provides. But consumers can define their own, too.

So, one thought is that we could provide an `Other(Box<dyn Error + Send + 'static>)` variant for user-implemented extractors? I hadn't realized that the extractor traits were meant to be publicly implemented.

> Also, it seems like a lot of the actual error cases probably overlap (e.g., you can have "invalid input" in a typed body, path param, query param, etc. and it's not clear a consumer should have to handle these all separately).

So, I also wondered about changing these to be more about the semantic error than the specific extractor that generated it. This approach is partially because some of the extractor errors represent things that have recommended HTTP semantics (e.g. a missing or invalid `Content-Type` header is supposed to return a 415 Unsupported Media Type status). But, I think we can just special-case those as part of a single `ExtractorError` that's shared across all extractors. I'll rework this a bit!

Reply:
> > I'm not sure about this approach here. These are just the extractors that Dropshot provides. But consumers can define their own, too.
> 
> So, one thought is that we could provide an `Other(Box<dyn Error + Send + 'static>)` variant for user-implemented extractors? I hadn't realized that the extractor traits were meant to be publicly implemented.

That seems like it would technically work but feels like second-classing them a bit.  The two things I mentioned combined made me think that organizing these variants by the type of extractor isn't great.  That makes me want to go down the other path you mentioned, organizing this more semantically.  But did you go down that path and find it problematic, too?

Reply:
Not exactly, I think organizing the error type variants semantically is fine. What I meant is just that I think it's important that _some_ extractor errors have first class representations (e.g. being enum variants that are easy to match on) because they have specific recommended response status codes/headers. This includes content-type errors and the websocket errors. Basically everything else can probably be grouped into a generic 400 Bad Request + message, or a generic 500 Internal Server Error.

Reply:
I think the real question becomes: "beyond those errors which we know ought to have specific response statuses etc, what categories of errors might users want to programmatically match on?". Pretty much every error that _we_ return that isn't content type related is "some kind of parse error", so we could have an `InvalidInput` variant or something, and perhaps there should be an internal error variant for "the extractor is broken".  I'm on the fence as to whether those variants ought to be represented by `Box<dyn Error>` or by dropshot's `HttpError`...

Reply:
For the Dropshot-provided errors, I think the two use cases are:

- programmatic consumers will treat them either like generic network-like errors (similar to a 500) or an unhandleable error that shouldn't happen (like a 400).  Their real goal from this work is being able to get a typed error for the errors returned by handlers, not the Dropshot ones.
- the other main consumer is people who want to serve a custom web page to a browser (@jclulow wants this and I think @david-crespo has too).  I'm not sure if they need much programmatic information about what went wrong.  (They probably _do_ want, like, the URL that was requested, though -- they might want to redirect to something that depends on it.)

Maybe the way to go is to start pretty generic (similar to `HttpError`: status code, internal message, external message) and extend it based on what people then ask for.

Reply:
Yeah, that makes sense; I'm inclined to potentially still represent it as an enum internally because I'd like some of the error variants to also be able to provide a "recommended headers" interface lazily, but I think it seems reasonable to wrap an enum behind an opaque type that just says "here's my recommended status code, internal/external message, and a method that adds headers associated with the error to a `HeaderMap`.